flowmachine/flowmachine/features/subscriber/event_count.py

```diff
@@ -99,8 +102,258 @@ def _make_query(self):
+                time_col="datetime",
+                size=size,
+                polygon_table=polygon_table,
+                geom_col=geom_col,
+            )
+
+        super().__init__()
+
+    @property
+    def column_names(self):
+        return ["subscriber", "value"]
+
+    def _make_query(self):
+        loc_cols = ", ".join(get_columns_for_level(self.level, self.column_name))
+
+        where_clause = ""
+        if self.direction != "both":
+            where_clause = (
+                f"WHERE outgoing IS {'TRUE' if self.direction == 'out' else 'FALSE'}"
+            )
+
+        return f"""
+        SELECT subscriber, {self.statistic}(events) AS value
+        FROM (
+            SELECT subscriber, {loc_cols}, COUNT(*) AS events
+            FROM ({self.unioned_query.get_query()}) U
+            {where_clause}
+            GROUP BY subscriber, {loc_cols}
+        ) U
+        GROUP BY subscriber
+        """
+
+
+class PerContactEventCount(SubscriberFeature):
+    """
+    This class returns the statistics of event count per contact per
+    subscriber within the period, optionally limited to only incoming or
+    outgoing events. For instance, it calculates the average number of events
+    per contact per subscriber.
+
+    Parameters
+    ----------
+    start, stop : str
+         iso-format start and stop datetimes
+    statistic : {'count', 'sum', 'avg', 'max', 'min', 'median', 'mode', 'stddev', 'variance'}, default 'avg'
+        Defaults to avg, aggregation statistic over the durations.
+    hours : 2-tuple of floats, default 'all'
+        Restrict the analysis to only a certain set
+        of hours within each day.
+    subscriber_identifier : {'msisdn', 'imei'}, default 'msisdn'
+        Either msisdn, or imei, the column that identifies the subscriber.
+    subscriber_subset : str, list, flowmachine.core.Query, flowmachine.core.Table, default None
+        If provided, string or list of string which are msisdn or imeis to limit
+        results to; or, a query or table which has a column with a name matching
+        subscriber_identifier (typically, msisdn), to limit results to.
+    direction : {'in', 'out', 'both'}, default 'out'
+        Whether to consider calls made, received, or both. Defaults to 'out'.
+    tables : str or list of strings, default 'all'
+        Can be a string of a single table (with the schema)
+        or a list of these. The keyword all is to select all
+        subscriber tables
+
+    Examples
+    --------
+
+    >>> s = PerContactEventCount("2016-01-01", "2016-01-07")
+    >>> s.get_dataframe()
+
+          subscriber      value
+    J0Yyqw2rkVEwpMG2       13.5
+    xkZb5E55LYE10wa4        9.5
+    oqNR8gkbv6e4K97z        9.5
+    2GJxeNazvlgZbqj6        9.5
+    D6b8NwVBmmw5JzA1       11.5
+                 ...        ...
+    """
+
+    def __init__(
+        self,
+        start,
+        stop,
+        contact_balance,
+        statistic="avg",
+        *,
+        hours="all",
```

Comment:
Can we get rid of these please - they now don't do anything.

Reply:
Of course!